### PR TITLE
Upgrading search options

### PR DESCRIPTION
**Pull request by itsellej**

### CODE REVIEW

File: lib/contact_manager.rb [hunk 1]
```diff
@@ -59,9 +59,35 @@ def begin
...
+          puts "Search by:"
+          puts "1. First name"
+          puts "2. Last name"
+          puts "3. Email address"
+          puts "4. Phone number"
+
+          selection = $stdin.gets.chomp
+
+          case selection
+          when "1"
```
Comment (andrewMacmurray) [inline lib/contact_manager.rb]: I like this 👍  very clean and readable

File: spec/contact_spec.rb [hunk 2]
```diff
@@ -1,34 +1,34 @@
 require 'contact'
 
 RSpec.describe Contact do
-  it "creates a new instance of the Contact class" do
+  xit "creates a new instance of the Contact class" do
```
Comment (andrewMacmurray) [inline spec/contact_spec.rb]: Any reason these are ignored?

File: spec/contact_spec.rb [hunk 3]
```diff
@@ -1,34 +1,34 @@
...
+  xit "create_new_entry method accepts input, creates new instance of Person and adds to @contacts" do
     allow($stdin).to receive(:gets).and_return("Elle", "H", "elle@hello.com", "07999999999")
     contacts = Contact.new
     contacts.create_new_entry
     expect(contacts.contacts).to eq([{"first_name" => "Elle", "last_name" => "H", "email_address" => "elle@hello.com", "phone_number" => "07999999999"}])
   end
 
-  it "create_new_entry method accepts input, creates new instance of Person and adds to @contacts twice" do
+  xit "create_new_entry method accepts input, creates new instance of Person and adds to @contacts twice" do
     allow($stdin).to receive(:gets).and_return("Elle", "H", "elle@hello.com", "07999999999", "Deneice", "Daniels", "dee@hello.com", "07889999999")
```
Comment (andrewMacmurray) [inline spec/contact_spec.rb]: There's a way you can reuse this logic for each of your tests and not have to repeat yourself, with a `before` block. Before your first test you can put something like:

```ruby
before(:each) do
   allow($stdin).to receive(:gets).and_return("Elle", "H", "elle@hello.com", "07999999999")
end
```

File: spec/contact_spec.rb (hunk 2 above)
Comment (itsellej) [reply spec/contact_spec.rb]: Yes, when I run a test it adds the info to the .json file. When the test after it is run, there's information it in from the previous test. As a result the output isn't as expected and the test fails. I hope I've explained it clearly :)

File: spec/contact_spec.rb (hunk 3 above)
Comment (itsellej) [reply spec/contact_spec.rb]: Thank you 👍

File: lib/contact.rb [hunk 4]
```diff
@@ -88,6 +88,75 @@ def search_last_name(search_term)
...
+      puts ""
+    end
+    puts "Please select which contact you wish to delete (select a number)"
+
+    selection = $stdin.gets.chomp.to_i
+
+    while (1..@contacts.length).to_a.include?(selection) == false
+      puts "You've entered an invalid selection. Please try again:"
+      selection = $stdin.gets.chomp.to_i
+    end
```
Comment (andrewMacmurray) [inline lib/contact.rb]: This is more of a style thing but I quite like to put blocks of code like this in helper methods, for example you could write:

```ruby
def get_selection
    selection = $stdin.gets.chomp.to_i
    while (1..@contacts.length).to_a.include?(selection) == false
      puts "You've entered an invalid selection. Please try again:"
      selection = $stdin.gets.chomp.to_i
    end
    selection
end
```

And then you can use that in `delete_contact`

```ruby
selection = get_selection
```

A good practice is also to keep these helper methods private (so others classes that use `Contact` can't call them by accident). You can do this by declaring these methods near the bottom of the class under the `private` keyword, for example:

```ruby
class Contact
   def delete_contact
       ...
   end

   private

   def get_selection
       ...
   end
end
```

Let me know if that doesn't make sense!

Comment (itsellej) [reply lib/contact.rb]: This is really helpful, thank you. I'll make this change now

File: spec/contact_spec.rb (hunk 3 above)
Comment (itsellej) [reply spec/contact_spec.rb]: For some reason, this doesn't seem to work when I call a method multiple times, like this:

`my_contacts = Contact.new
    my_contacts.create_new_entry
    my_contacts.create_new_entry
    my_contacts.create_new_entry
    expect(my_contacts.search_phone)...`

I can't seem to figure out why. Do you have any idea?

File: spec/contact_spec.rb (hunk 2 above)
Comment (andrewMacmurray) [reply spec/contact_spec.rb]: Ah, so you need to mock out the code that's writing to the file then 😉 

something like:

```ruby
allow(File).to receive(:write).with(your_data)
```

Comment (itsellej) [reply spec/contact_spec.rb]: I'll give this a go, thanks Andrew!

**Pull request merged**